### PR TITLE
changes cisco_fxos_sensors to base on entity_sensors (reopen PR)

### PR DESCRIPTION
**Pull request by thl-cmk**
-> remove *fxos* files
-> split entity_sensors.py in utils/enttity_sensors.py end entity_sensors.py
-> new cisco_entity_sensors.py: register.snmp/parse function (not for undefined/temp) -> checks in entity_sensors.py
-> new test_cisco_entity_sensors.py

Known Issues:
-> Service name for power_presence "Power Sensor" overlaps with power_consumption sensors -> change to "Power supply"
-> default values lower for fan are in conflict with device status -> needs options like 'temperature'
-> some ASA devices have no valid scaling value (always 1), posible workaround:
   in "parse_entity_sensors" check scaling value -> if 1 change to 9

Posible improvements:
-> WATO: hw_fan.py -> change "output_metrics" -> CheckBox to FixedValue -> on click less ;-)
-> name field in EntitySensor is not used -> can be removed

### CODE REVIEW

File: cmk/base/plugins/agent_based/entity_sensors.py [hunk 1]
```diff
@@ -184,9 +114,9 @@ def discover_entity_sensors_temp(section: EntitySensorSection) -> DiscoveryResul
 
 
 def check_entity_sensors_temp(
-    item: str,
-    params: TempParamType,
-    section: EntitySensorSection,
+        item: str,
```
Comment (jherbel) [inline cmk/base/plugins/agent_based/entity_sensors.py:117]: Please undo these changes

File: cmk/base/plugins/agent_based/cisco_entity_sensors.py [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+def parse_cisco_entity_sensors(string_table: List[StringTable]) -> EntitySensorSection:
+    section: EntitySensorSection = {}
+    sensor_names = {i[0]: i[1] for i in string_table[0]}
+    for oid_end, sensor_type_nr, scaling_nr, reading, status_nr, device_unit in string_table[1]:
+        # Some devices support the ENTITY-MIB including sensor/entity names.
+        # Others (e.g. Cisco Firepower FPR-1140 Security Appliance) do not support
+        # this MIB, thus we use OID as item instead
+
+        # do not add undefined and temp sensors (duplicate with cisco_temperature)
+        if sensor_type_nr not in ['0', '8']:
```
Comment (jherbel) [inline cmk/base/plugins/agent_based/cisco_entity_sensors.py:94]: We have quite some duplicate code here. It would be better to define a generic parse function in utils/entity_sensors.py which takes an additional input argument 
`sensor_types_ignore: Container[str]`
You can then use this function for both snmp sections, for example
```
def parse_entity_sensors(string_table: List[StringTable]):
    return utils.parse_entity_sensors(string_table, set())
```

**Review by jherbel — CHANGES_REQUESTED** [review]
Besides some minor issues, we should discuss the known issues you mentioned:

- Service name for power_presence "Power Sensor" overlaps with power_consumption sensors -> change to "Power supply"
Where exactly does this overlap occur? Within this plugin or are there overlaps with another plugin?
- default values lower for fan are in conflict with device status -> needs options like 'temperature'
Can this for the moment be overcome by changing the lower levels with a rule?
- some ASA devices have no valid scaling value (always 1), posible workaround:
in "parse_entity_sensors" check scaling value -> if 1 change to 9
What value do we get from entPhySensorScale report in such cases?

**Comment by thl-cmk** [comment]
> Besides some minor issues, we should discuss the known issues you mentioned:
> 
> * Service name for power_presence "Power Sensor" overlaps with power_consumption sensors -> change to "Power supply"
>   Where exactly does this overlap occur? Within this plugin or are there overlaps with another plugin?
true, to a other plugin, at the moment it's a CheckPoint/Dell Server, this has a Power supply presence/function sensor and a Power consumtion sensor, so in the long run it would make sense to have a common naming over all these type of plugins. Might be better to change it now before the plugin is out with the next public release.

> * default values lower for fan are in conflict with device status -> needs options like 'temperature'
>   Can this for the moment be overcome by changing the lower levels with a rule? 
it could, but it's kind of ugly

> * some ASA devices have no valid scaling value (always 1), posible workaround:
>   in "parse_entity_sensors" check scaling value -> if 1 change to 9
>   What value do we get from entPhySensorScale report in such cases? 
The device reports 1, meaning 10**(-24)

File: tests/unit/cmk/base/plugins/agent_based/test_cisco_entity_sensors.py [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+    EntitySensor,
+)
+from cmk.base.plugins.agent_based.agent_based_api.v1 import (
+    State,
+    Service,
+    Result,
+)
+
+from cmk.base.plugins.agent_based.entity_sensors import (
+    check_entity_sensors_fan,
```
[the remark this answers is not in the document]
Comment (thl-cmk) [reply tests/unit/cmk/base/plugins/agent_based/test_cisco_entity_sensors.py:23]: > Any tests for entity_sensors should be implemented in test_entity_sensors.py. Here, you should only test the parse function.

Ok, so i move the test over to test_entity_sensors.py (except for the parse section), or is there than no additional testing nessecary? And I can remove these additional tests?

[the remark this answers is not in the document]
Comment (jherbel) [reply tests/unit/cmk/base/plugins/agent_based/test_cisco_entity_sensors.py:23]: You can move them over

**Comment by jherbel** [comment]
> true, to a other plugin, at the moment it's a CheckPoint/Dell Server, this has a Power supply presence/function sensor and a Power consumtion sensor, so in the long run it would make sense to have a common naming over all these type of plugins. Might be better to change it now before the plugin is out with the next public release.
which plugin?

> The device reports 1, meaning 10**(-24)
Can 1 ever be correct? Do we have a way of knowing?

**Comment by thl-cmk** [comment]
> > true, to a other plugin, at the moment it's a CheckPoint/Dell Server, this has a Power supply presence/function sensor and a Power consumtion sensor, so in the long run it would make sense to have a common naming over all these type of plugins. Might be better to change it now before the plugin is out with the next public release.
> 
> which plugin?  
checkpoint_sensors, not yet public ;-)

> 
> > The device reports 1, meaning 10**(-24)
> 
> Can 1 ever be correct?
for fan speed I think 1 can't never be correct or it must be a very slow fan
for temp, I think also not, at least for a sensor
>  Do we have a way of knowing?
I think not realy. In my option the best way would be to have some kind of option for the parse function, so this could become configurable.

From the MIB description of sensorValue: "Range: -1000000000 - 1000000000"
so a fan with sensorValue = 1000000000 (max) and sensorScale =1 (as reported) will rotate at 9.999999999999999e-16 RPM
this must be a kind of very acurate sensor to measure this little value, I think this is true for temp also

**Comment by jherbel** [comment]
> checkpoint_sensors, not yet public ;-)

Ok, in that case, everything is fine I would say, since its not even a plugin which is part of Checkmk

> From the MIB description of sensorValue: "Range: -1000000000 - 1000000000"
so a fan with sensorValue = 1000000000 (max) and sensorScale =1 (as reported) will rotate at 9.999999999999999e-16 RPM
this must be a kind of very acurate sensor to measure this little value, I think this is true for temp also

Then it would probably make sense to simply hard-code this. I.e., neither temperature nor fan sensors should ever have sensorScale =1 and if this occurs, we set sensorScale =9. However, I would suggest do this afterwards in a separate PR afterwards (small bugfix).

**Comment by thl-cmk** [comment]
> > checkpoint_sensors, not yet public ;-)
> 
> Ok, in that case, everything is fine I would say, since its not even a plugin which is part of Checkmk

This is not entirely true, these checks are already there (except for the Power consumption) but sitll in /checks.
Especially checkpoint_powersupply, where the service name is "Power Supply". Basically this is the same as the entity_sensors_power check. Just curious, why is this such a problem to straighten the names?

> 
> > From the MIB description of sensorValue: "Range: -1000000000 - 1000000000"
> > so a fan with sensorValue = 1000000000 (max) and sensorScale =1 (as reported) will rotate at 9.999999999999999e-16 RPM
> > this must be a kind of very acurate sensor to measure this little value, I think this is true for temp also
> 
> Then it would probably make sense to simply hard-code this. I.e., neither temperature nor fan sensors should ever have sensorScale =1 and if this occurs, we set sensorScale =9. However, I would suggest do this afterwards in a separate PR afterwards (small bugfix).

Ok. could look like this in the parse function:
```
if sensor_type_nr in ["8", "10"] and scaling_nr == "1":
            scaling_nr = "9"
```

File: cmk/base/plugins/agent_based/cisco_entity_sensors.py [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+                state=sensor_state(status_nr),
+                status_descr=sensor_status_descr(status_nr),
+            )
+    return section
+
+
+register.snmp_section(
+    name='cisco_entity_sensors',
+    detect=all_of(startswith(OIDSysDescr, 'Cisco Firepower'),
+                  contains(OIDSysDescr, 'security appliance'),
```
[the remark this answers is not in the document]
Comment (thl-cmk) [reply cmk/base/plugins/agent_based/cisco_entity_sensors.py:110]: I don't think the are in conflict. From cisco_entity_sensors
```
    detect=all_of(startswith(OIDSysDescr, 'Cisco Firepower'),
                  contains(OIDSysDescr, 'security appliance'),
                  ),
```

and from  entity_sensors
```
    detect=any_of(
        startswith(OIDSysDescr, "palo alto networks"),
        startswith(OIDSysDescr, "cisco adaptive security appliance"),
    ),
```

In cisco_entity_sensors there needs to be a match on both conditions (basically "Cisco Firepower <some text> security appliance" and in entity_sensors there needs to be a match on "cisco adaptive security appliance". So if we dont have a "Cisco Firepower cisco adaptive security appliance" we are fine I think.

[the remark this answers is not in the document]
Comment (jherbel) [reply cmk/base/plugins/agent_based/cisco_entity_sensors.py:110]: Basically that is my questions, so there are no things like "Cisco Firepower security appliance"?

**Comment by jherbel** [comment]
> This is not entirely true, these checks are already there (except for the Power consumption) but sitll in /checks.
Especially checkpoint_powersupply, where the service name is "Power Supply". Basically this is the same as the entity_sensors_power check. Just curious, why is this such a problem to straighten the names?

Ah sry, I thought you meant check plugins that you had lying around locally or something. In general, changing service descriptions is problematic, since it can affect a lot of customers (you have to maybe changes rules etc.). I don't see the overlap between checkpoint_powersupply and entity_sensors. checkpoint_powersupply has
'service_description': "Power Supply %s",
and entity_sensors_power_presence has
service_name='Power %s',
However, in entity_sensors, we always have "Sensor" in the item name (_reformat_sensor_name).

> Ok. could look like this in the parse function:
if sensor_type_nr in ["8", "10"] and scaling_nr == "1":
            scaling_nr = "9"

Yes, but as I said, please in a separate commit afterwards.

**Comment by thl-cmk** [comment]
> Ah sry, I thought you meant check plugins that you had lying around locally or something.....
I understand that, on the other hand, it is very annoying when you have products from different vendors and the "basically" same thing in CMK has different functions and names. In this case, for example, I couldn't just say that all power sensors are shown globally because they have different names. The scond thing, this check is very new, meaning not a lot of custumers will have this yet. So changing it now will be more easy than later when every one is using it.  But in the end your decission.

One more thing about the naming. A quick grep over your code base shows, that most of this checks use the term "Power supply" (18), some use "Powersupply" (2) or "Power supply satus" (2). and there is at least one "Power consumtion" check  (fsc_sc2_power_consumption).

**Comment by jherbel** [comment]
Ok, what exactly do you want to rename into what?

**Comment by thl-cmk** [comment]
From:
```
    service_name='Power %s',
```

to:
```
    service_name='Power Supply %s',
```

**Comment by jherbel** [comment]
For now, we leave this as it is.

File: cmk/base/plugins/agent_based/cisco_entity_sensors.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (thl-cmk) [reply cmk/base/plugins/agent_based/cisco_entity_sensors.py:110]: no, not to my  knowledge.

File: tests/unit/cmk/base/plugins/agent_based/test_cisco_entity_sensors.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (thl-cmk) [reply tests/unit/cmk/base/plugins/agent_based/test_cisco_entity_sensors.py:23]: OK. If I move the discover function tests we will still use functions from both plugins. Is this ok.
e.g. "discover_entity_sensors_power_presence" and "parse_cisco_entity_sensors" in "test_discover_entity_sensors_power_presence"

[the remark this answers is not in the document]
Comment (jherbel) [reply tests/unit/cmk/base/plugins/agent_based/test_cisco_entity_sensors.py:23]: No, its better to not mix that. Just copy _section over to the other file. Also, I guess the discovery and the check function are well tested, so there is no need to add that many tests.

File: cmk/base/plugins/agent_based/cisco_entity_sensors.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (jherbel) [reply cmk/base/plugins/agent_based/cisco_entity_sensors.py:110]: Ok then this is ok

File: cmk/base/plugins/agent_based/utils/entity_sensors.py [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+    '''Converts device units to units known by Check_mk'''
+    return {
+        'celsius': 'c',
+        'fahrenheit': 'f',
+        'kelvin': 'k',
+    }.get(unit)
+
+
+def parse_entity_sensors(string_table: List[StringTable],
+                         sensor_types_ignore: Optional[Container[str]] = None) -> EntitySensorSection:
```
Comment (jherbel) [inline cmk/base/plugins/agent_based/utils/entity_sensors.py:100]: you can use `set()` as default here

File: cmk/base/plugins/agent_based/utils/entity_sensors.py [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+    '''
+    :param string_table:
+    :param sensor_types_ignore: list of sensor types to ignore. e.g ['0', '8'] for undefined and temperature
+    '''
+    if sensor_types_ignore is None:
+        sensor_types_ignore = []
+    section: EntitySensorSection = {}
+    sensor_names = {i[0]: i[1] for i in string_table[0]}
+    for oid_end, sensor_type_nr, scaling_nr, reading, status_nr, device_unit in string_table[1]:
+        if sensor_type_nr not in sensor_types_ignore:
```
Comment (jherbel) [inline cmk/base/plugins/agent_based/utils/entity_sensors.py:110]:
```
if sensor_type_nr in sensor_types_ignore: 
    continue
```
this avoid indenting the whole block afterwards

File: cmk/base/plugins/agent_based/utils/entity_sensors.py (hunk 5 above)
Comment (jherbel) [inline cmk/base/plugins/agent_based/utils/entity_sensors.py:100]: use `set()` here (and typization `Container[str]`)

**Review by jherbel — CHANGES_REQUESTED** [review]
The unit tests for the parse function can all be moved to .../utils/test_entity_sensors.py. There, you can tests cases with and without exclusion.

**Comment by thl-cmk** [comment]
> The unit tests for the parse function can all be moved to .../utils/test_entity_sensors.py. There, you can tests cases with and without exclusion.

You mean the last test in .../utils/test_cisco_entity_sensors.py? Even if then have a cross reference betwen entity_sensors and cisco_entity_sensors? If yes, I will then remove .../utils/test_cisco_entity_sensors.py completely.

**Comment by jherbel** [comment]
> You mean the last test in .../utils/test_cisco_entity_sensors.py? Even if then have a cross reference between entity_sensors and cisco_entity_sensors? If yes, I will then remove .../utils/test_cisco_entity_sensors.py completely.

What I mean is that there is no need to have separate tests for `parse_cisco_entity_sensors` and `entity_sensors.parse_entity_sensors`. Simply test the parse function `utils.entity_sensors`.

**Review by jherbel — APPROVED** [review]
Looks good from my side, are there any other changes you want to do here? Otherwise Ill go ahead and merge.

**Comment by thl-cmk** [comment]
Hi Joerg,

I think for now we should move forward with this PR. The oher things we discussed, I guess we need to do this at a later time.

/Thomas

**Comment by jherbel** [comment]
Ok, then I will go ahead and merge this. Bugfixes (such as the scaling) can be handled in later PRs.
